solr/core/src/test/org/apache/solr/store/shared/SharedCoreConcurrencyTest.java

```diff
@@ -49,411 +58,673 @@
+
+    Replica newReplica = null;
+
+    for (Replica r : getCollection().getSlice(SHARD_NAME).getReplicas()) {
+      if (!existingReplicas.contains(r.getName())) {
+        newReplica = r;
+        break;
+      }
     }
+
+    assertNotNull("Could not find new replica", newReplica);
+
+    return newReplica;
   }
 
-  private void configureTestSharedConcurrencyControllerForNode(JettySolrRunner runner, List<String> progress) {
-    SharedCoreConcurrencyController concurrencyController = new SharedCoreConcurrencyController(runner.getCoreContainer()) {
-      Object recorderLock = new Object();
-      @Override
-      public void recordState(String collectionName, String shardName, String coreName, SharedCoreStage stage) {
-        super.recordState(collectionName, shardName, coreName, stage);
+  /**
+   * Directly query the {@code replica} and verify that all the docs indexed(after accounting for deletions) are found.
+   */
+  private void queryNewReplicaAndVerifyAllDocsFound(Replica replica) throws Exception {
+    try (SolrClient replicaDirectClient = getHttpSolrClient(replica.getBaseUrl() + "/" + replica.getCoreName())) {
+      ModifiableSolrParams params = new ModifiableSolrParams();
+      params.set("q", "*:*").set("distrib", "false").set("rows", testState.numDocsIndexed.get());
+      CountDownLatch latch = new CountDownLatch(1);
+      Map<String, CountDownLatch> corePullTracker = testState.getCorePullTracker(replica.getNodeName());
+      corePullTracker.put(replica.getCoreName(), latch);
+      QueryResponse resp = replicaDirectClient.query(params);
+      assertEquals("new replica did not return empty results", 0, resp.getResults().getNumFound());
 
-        synchronized (recorderLock) {
-          progress.add(Thread.currentThread().getId() + "." + stage.name());
-        }
+      assertTrue(latch.await(120, TimeUnit.SECONDS));
 
+      resp = replicaDirectClient.query(params);
+      List<String> docs = resp.getResults().stream().map(r -> (String) r.getFieldValue("id")).collect(Collectors.toList());
+      assertEquals("we did not ask for all the docs found", resp.getResults().getNumFound(), docs.size());
+      Collections.sort(docs, new Comparator<String>() {
+        public int compare(String id1, String id2) {
+          return Integer.parseInt(id1) - Integer.parseInt(id2);
+        }
+      });
+      List<String> docsExpected = new ArrayList<>();
+      for (int i = 1; i <= testState.numDocsIndexed.get(); i++) {
+        String docId = Integer.toString(i);
+        if (!testState.idsDeleted.contains(docId)) {
+          docsExpected.add(docId);
+        }
       }
-    };
-    setupTestSharedConcurrencyControllerForNode(concurrencyController, runner);
+      log.info("numDocsFound=" + docs.size() + " docsFound= " + docs.toString());
+      assertEquals("wrong docs", docsExpected.size() + docsExpected.toString(), docs.size() + docs.toString());
```

Comment:
docsExpected.size() + docsExpected.toString() intentional? Why not two separate assert statements

Reply:
Yes. It helps in debugging to see what docs are different. You can go in the earlier log lines and correlate to exact indexing batch.